src/utils/db.ts

```diff
@@ -0,0 +1,109 @@
+};
+
+// Initialize Firebase
+const app = initializeApp(firebaseConfig);
+const db = getFirestore(app);
+
+interface DialAttempt {
+  address: string;
+  peerId: string;
+  timestampOfDial: number;
+  connectionStatus: boolean;
+}
+
+interface UptimeStats {
+  dialAttempts: number;
+  successes: number;
+  uptimeRatio: number;
+}
+
+export const updateIncentivizedArchaeologists = async () => {
+  try {
+    console.log("Updating incentivizes archaeologists...", incentivizedArchaeologists.length);
+
+    for await (const address of incentivizedArchaeologists) {
+      await setDoc(doc(db, `incentivized_archaeologists/${address}`), { address });
+    }
+
+    console.log("DONE!");
+  } catch (e) {
+    console.error("Error adding document: ", e);
+  }
+};
+
+export const saveDialResults = (attempts: DialAttempt[], timestampOfDial: number) => {
+  try {
+    attempts.map(async record => {
+      addDoc(collection(db, "dial_attempts"), record);
+    });
+
+    addDoc(collection(db, "dial_times"), { time: timestampOfDial });
+  } catch (e) {
+    console.error("Error adding document: ", e);
+  }
+};
+
+export const getOnlineNodes = async () => {
+  try {
+    const dialTimes = await getDocs(query(collection(db, "dial_times"), orderBy("time", "desc")));
+    const lastDialTime = dialTimes.docs[0].get("time");
+
+    const latestOnlineNodesSnapshot = await getDocs(
+      query(
+        collection(db, "dial_attempts"),
+        where("timestampOfDial", "==", lastDialTime),
+        where("connectionStatus", "==", true)
+      )
+    );
+
+    const onlinePeerIds: string[] = latestOnlineNodesSnapshot.docs.map(node => node.get("peerId"));
+    return onlinePeerIds;
+  } catch (e) {
+    console.error("Error adding document: ", e);
+    throw e;
+  }
+};
+
+export const getUptimeStats = async (fromTimestamp?: number) => {
+  try {
+    const incentivizedArchsSnapshot = await getDocs(collection(db, "incentivized_archaeologists"));
+    const incentivizedArchs: string[] = incentivizedArchsSnapshot.docs.map(doc => doc.get("address"));
+
+    const dialTimes = await getDocs(query(collection(db, "dial_times"), orderBy("time", "desc")));
+
+    const nDialAttempts = fromTimestamp
+      ? dialTimes.docs.filter(t => t.get("time") >= fromTimestamp).length
+      : dialTimes.docs.length;
+
+    const uptimeStatistics: Record<string, UptimeStats> = {};
+
+    const successfulDialsSnapshot = (
+      await getDocs(query(collection(db, "dial_attempts"), where("connectionStatus", "==", true)))
+    ).docs;
+
+    for await (const archAddress of incentivizedArchs) {
+      const successes = successfulDialsSnapshot.filter(
+        dial =>
+          dial.get("address") === archAddress && (fromTimestamp ? dial.get("timestampOfDial") >= fromTimestamp : true)
+      ).length;
+
+      uptimeStatistics[archAddress] = {
+        dialAttempts: nDialAttempts,
+        successes,
+        uptimeRatio: successes / nDialAttempts,
+      };
+    }
```

Comment:
I didn't realize firebase doesn't have aggregation queries, which would be perfect for this scenario b/c they would let you spit out aggregate totals (i.e. "address", "count of successes", "count of failures"). B/c we can't do this, it makes sense that there's so many read operations since we have to return the entire collection.

To avoid the read-amount issue, what if instead we keep a running total of successes and failures every time we do the writes to the database?

So in this instance, we would:

1. Add 2 more columns to the arch collection -- `success_count` + `fail_count` (or whatever you want to call them)
2. When dialing, increment the appropriate column for that archaeologist based on success/failure

Then we would just have 1 record per archaeologist in the collection (rather than a new record for each dial attempt).

The uptime ratio could then be `total_successes / (total_failures + total_successes)`. But we could still keep track of the dial times to detect the dial index at which an archaeologist started being dialed - `total_dials - (total_failures + total_successes)`

 And then the API call to populate the front-end would just return the set of archaeologists that have their `connectionStatus` set to true (and we wouldn't need to take into account the dial time).

Let me know if this does or doesn't make sense.

Reply:
@DarksightKellar -- good point about losing the timing of when dials failed.

I think we can make this tradeoff for now (saving reads at the expense of losing the filter). Think its good to move forward with the extra columns for aggregation update.